### PR TITLE
slayer: Modify to use RSProfiles Config

### LINKED ISSUE #249: Slayer plugin - Task carrying over to other accounts
**Detailed Problem:**
Slayer task amount is only saved for the client, not per account. As a result, when a user switches accounts it will still show the slayer task from the previous account, and even progress the task if not checked on the new account.
**Cause/Limitations:**
Slayer task amount is saved as an extra variable in the saved plugin settings. There is no distinction between different accounts, because plugins settings are global between all accounts, Plugin settings currently do not support saving Maps because the database is not structured to handle them.

### PR DESCRIPTION
**Pull request by Cyborger1**
Modifies the Slayer Plugin to use the newly added RSProfiles feature to ConfigManager. This closes #249.

The website actually pulls the info from the old config fields though.
I'm not sure what needs to be done on that end. Someone who knows his way around the rl.net code would have to help with this.
With that said, this shouldn't be merged until the website can support the new config path.

### CODE REVIEW

File: runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java [hunk 1]
```diff
@@ -919,7 +932,8 @@ public void testNewAccountSlayerKill()
 	@Test
 	public void infoboxNotAddedOnLogin()
 	{
-		when(slayerConfig.taskName()).thenReturn(Task.BLOODVELD.getName());
+		when(slayerPlugin.getStoredTaskName()).thenReturn(Task.BLOODVELD.getName());
+		when(slayerPlugin.getStoredAmount()).thenReturn(50);
```
Comment (Nightfirecat) [inline runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java:936]: Why is the amount set in this test?

File: runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java [hunk 2]
```diff
@@ -283,7 +291,8 @@ public void testBossTask()
 
 		assertEquals("Vet'ion", slayerPlugin.getTaskName());
 		assertEquals(3, slayerPlugin.getAmount());
-		verify(slayerConfig).points(914);
+
```
Comment (Nightfirecat) [inline runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java:294]: These added newlines in this file can be removed

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 3]
```diff
@@ -275,27 +292,132 @@ public void onGameStateChanged(GameStateChanged event)
...
+		setIntConfig(SLAUGHTER_CHARGES_KEY, slaughterCharges);
+	}
+
+	private int getIntConfig(String key)
+	{
+		Integer value = configManager.getRSProfileConfiguration(SlayerConfig.GROUP_NAME, key, int.class);
+		return value == null ? -1 : value;
+	}
+
+	private void setIntConfig(String key, int value)
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java:397]: This is identical to `setStringConfig` except for the signature. Change the signature to `(String key, Object value)` and direct all setters to use it.

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 4]
```diff
@@ -275,27 +292,132 @@ public void onGameStateChanged(GameStateChanged event)
...
+						getStoredAmount(),
+						getStoredInitialAmount(),
+						getStoredTaskLocation(), false);
 					loginFlag = false;
 				}
 				break;
 		}
 	}
 
+	public int getStoredAmount()
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java:311]: All getters should be package-private, and setters should be private.

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 5]
```diff
@@ -133,12 +133,25 @@
...
 
+	// Key names for stored task values
+	public static final String TASK_NAME_KEY = "taskName";
+	public static final String AMOUNT_KEY = "amount";
+	public static final String INIT_AMOUNT_KEY = "initialAmount";
+	public static final String TASK_LOC_KEY = "taskLocation";
+	public static final String STREAK_KEY = "streak";
+	public static final String POINTS_KEY = "points";
+	public static final String EXPEDITIOUS_CHARGES_KEY = "expeditious";
+	public static final String SLAUGHTER_CHARGES_KEY = "slaughter";
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java:144]: These should be package-private and `@VisibleForTesting`

**Review by Nightfirecat — CHANGES_REQUESTED** [review]

File: runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java (hunk 1 above)
Comment (Cyborger1) [reply runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java:936]: I figured since this tests [this code branch](https://github.com/runelite/runelite/blob/a58e441f9a236eff8b2012761189f1fe4b6f9db7/runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java#L295) I would set an arbitrary amount value along with the name. If this defaults to 0 otherwise I can remove it.

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java (hunk 5 above)
Comment (Cyborger1) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java:144]: Thank you, I forgot about that annotation.

File: runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java (hunk 1 above)
Comment (Cyborger1) [reply runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java:931]: I fiddled around with this some more. The amount check, and `when(slayerConfig.showInfobox()).thenReturn(true);`, which I'll be adding, are both required to properly test this behavior.

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 6]
```diff
@@ -133,12 +133,33 @@
 	private static final Pattern TASK_STRING_VALIDATION = Pattern.compile("[^a-zA-Z0-9' -]");
 	private static final int TASK_STRING_MAX_LENGTH = 50;
 
+	// Key names for stored task values
+	@VisibleForTesting
+	static final String TASK_NAME_KEY = "taskName";
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java:138]: these should go in the Config

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 7]
```diff
@@ -275,27 +300,127 @@ public void onGameStateChanged(GameStateChanged event)
...
+						getStoredAmount(),
+						getStoredInitialAmount(),
+						getStoredTaskLocation(), false);
 					loginFlag = false;
 				}
 				break;
 		}
 	}
 
+	int getStoredAmount()
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java:319]: naming these like getters/setters is somewhat confusing. Its probably clearest for them to be replaced with calls directly to configmanager

Comment (Cyborger1) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java:319]: I agree we could change the name, but I'm not sure replacing these with direct calls to the manager would be clearer. The call can get quite long to type out and the getters abstract away a null check to keep the behavior consistent with the old config setup.

Comment (abextm) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java:301]: Just `getIntConfig(SlayerConfig.WHATEVER)` seems fine as it resolves the naming problem.

Comment (Cyborger1) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java]: Thank god for ctrl-alt-n

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java [hunk 8]
```diff
@@ -275,27 +281,47 @@ public void onGameStateChanged(GameStateChanged event)
...
+						getIntProfileConfig(SlayerConfig.AMOUNT_KEY),
+						getIntProfileConfig(SlayerConfig.INIT_AMOUNT_KEY),
+						getStringProfileConfig(SlayerConfig.TASK_LOC_KEY), false);
 					loginFlag = false;
 				}
 				break;
 		}
 	}
 
+	int getIntProfileConfig(String key)
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java:300]: These should be `@VisibleForTesting` as they won't be used elsewhere.

**Review by Nightfirecat — COMMENTED** [review]
LGTM otherwise at a glance

File: runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java (hunk 8 above)
Comment (Cyborger1) [reply runelite-client/src/main/java/net/runelite/client/plugins/slayer/SlayerPlugin.java:301]: Done 👍

**Review by abextm — APPROVED** [review]

**Review by Nightfirecat — APPROVED** [review]

File: runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java [hunk 9]
```diff
@@ -994,4 +1006,13 @@ public void testMultikill()
 		assertEquals("Suqahs", slayerPlugin.getTaskName());
 		assertEquals(229, slayerPlugin.getAmount()); // 2 kills
 	}
+
+	// Helper method for ConfigManager mock
+	// Last call to the manager needs to have the expected value
+	private void verifyIntConfig(String key, Integer expectedValue)
+	{
+		ArgumentCaptor<Integer> captor = ArgumentCaptor.forClass(int.class);
+		verify(configManager, atLeastOnce()).setRSProfileConfiguration(eq(SlayerConfig.GROUP_NAME), eq(key), captor.capture());
```
Comment (Adam-) [inline runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java:1015]: Why was this changed to `atLeastOnce()`?

Comment (Cyborger1) [reply runelite-client/src/test/java/net/runelite/client/plugins/slayer/SlayerPluginTest.java]: In case the config gets set more than once, the only thing that matters is that the last call is of the expected value. Without `atLeastOnce()`, the captor serves no purpose and could just be a full equality check as the default behavior is that the call needs to happen exactly once.
